designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,326 @@
+
+  * Note: If `--help` flag is being passed from `kubebuilder` to the external plugin through `PluginRequest`, the plugin attaches its help text information in the `Help` field of the `PluginResponse`. Both `PluginRequest` and `PluginResponse` also contain `APIVersion` field to have compatible versioned schemas.
+
+* Handling plugin failures across the chain
+
+  * If any plugin in the chain fails, the plugin reports errors back through `PluginResponse` to `kubebuilder` and plugin chain execution will be halted, as one plugin may be dependent on the success of another. All the files that were scaffolded already until that point will not be written to disk to prevent a half committed state.
+
+## Implementation Details/Notes/Constraints
+
+```go
+// PluginRequest contains all information kubebuilder received from the CLI
+// and plugins executed before it.
+type PluginRequest struct {
+  // Command contains the command to be executed by the plugin such as init, create api, etc.
+  Command       string              `json:"command"`
+
+  // APIVersion defines the versioned schema of the PluginRequest that is encoded and sent from Kubebuilder to plugin.
+  // Initially, this will be marked as alpha (v1alpha1).
+  APIVersion    string              `json:"apiVersion"`
+
+  // Args holds the plugin specific arguments that are received from the CLI which are to be passed down to the plugin.
+  Args          []string            `json:"args"`
+
+  // Universe represents the modified file contents that gets updated over a series of plugin runs
+  // across the plugin chain. Initially, it starts out as empty.
+  Universe      map[string]string   `json:"universe"`
+}
+
+// PluginResponse is returned to kubebuilder by the plugin and contains all files
+// written by the plugin following a certain command.
+type PluginResponse struct {
+  // Command holds the command that gets executed by the plugin such as init, create api, etc.
+  Command       string                   `json:"command"`
+
+  // Help contains the plugin specific help text that the plugin returns to kubebuilder when it receives
+  // `--help` flag from Kubebuilder.
+  Help          string                   `json:"help,omitempty"`
+
+  // APIVersion defines the versioned schema of the PluginResponse that will be written back to kubebuilder.
+  // Initially, this will be marked as alpha (v1alpha1).
+  APIVersion    string                   `json:"apiVersion"`
+
+  // Universe in the PluginResponse represents the updated file contents that was written by the plugin.
+  Universe      map[string]string        `json:"universe"`
+
+  // Error is a boolean type that indicates whether there were any errors due to plugin failures.
+  Error         bool                     `json:"error,omitempty"`
+
+  // ErrorMsg holds the specific error message of plugin failures.
+  ErrorMsg      string                   `json:"error_msg,omitempty"`
```

Comment:
Is it just one error? Would we ever want to allow a series of errors to come back?

Reply:
No objection to making this `errList []error`. `errInfo map[string][]error` might be better where key is the plugin name. @estroz what do you think?

Reply:
Fine with me!